test/gtest/uct/ib/test_sockaddr.cc

```diff
@@ -548,4 +559,99 @@ UCS_TEST_P(test_uct_cm_sockaddr, cm_server_reject)
                 (TEST_CM_STATE_SERVER_CONNECTED | TEST_CM_STATE_CLIENT_CONNECTED)));
 }
 
+UCS_TEST_P(test_uct_cm_sockaddr, many_clients_to_one_server)
+{
+    int i, num_clients = 100;
+    entity *client_test;
+
+    UCS_TEST_MESSAGE << "Testing "     << m_listen_addr
+                     << " Interface: " << GetParam()->dev_name;
+
+    /* Listen */
+    cm_start_listen();
+
+    /* Connect */
+    /* multiple clients, each on a cm of its own, connecting to the same server */
+    for (i = 0; i < num_clients; ++i) {
+        client_test = uct_test::create_entity();
+        m_entities.push_back(client_test);
+        client_test->client_cb_arg = client_test->cm_attr().max_conn_priv;
+        client_test->connect(0, *m_server, 0, m_connect_addr,
+                             client_connect_cb, client_disconnect_cb, this);
+    }
+
+    /* wait for the server to connect to all the clients */
+    while ((m_client_connect_cb_cnt < num_clients) ||
+           (m_server_connect_cb_cnt < num_clients )) {
+        progress();
+    }
+    EXPECT_EQ(num_clients, m_server_recv_req_cnt);
+    EXPECT_EQ(num_clients, m_client_connect_cb_cnt);
+    EXPECT_EQ(num_clients, m_server_connect_cb_cnt);
+    EXPECT_EQ(num_clients, (int)m_server->num_eps());
+
+    /* Disconnect */
+    for (i = 0; i < num_clients; ++i) {
+        client_test = m_entities.back();
+        ASSERT_TRUE(client_test != m_client);
+        cm_disconnect(client_test);
+
+        /* don't remove the ep, i.e. don't call uct_ep_destroy before the client
+         * finished disconnecting so that a disconnect event won't arrive on a
+         * destroyed endpoint on the client side */
+        while (m_client_disconnect_cnt < (i + 1)) {
```

Comment:
pls move this loop outside of `for (i = 0; i < num_clients; ++i) {` otherwise it's serialized.

Reply:
but i can't remove this client before this loop ends.

Reply:
sure, remove this client later

Reply:
I mean remove all client after all of them are disconnected